core/src/main/scala/chisel3/Module.scala

```diff
@@ -283,6 +283,60 @@ abstract class Module extends RawModule {
   }
 }
 
+/** Provides an implicit Clock for use _within_ the [[RawModule]]
+  *
+  * Be careful to define the Clock value before trying to use it.
+  * Due to Scala initialization order, the actual val defining the Clock must occur before any
+  * uses of the implicit Clock.
+  *
+  * @example
+  * {{{
+  * class MyModule extends RawModule with ImplicitClock {
+  *   // Define a Clock value, it need not be called "implicitClock"
+  *   val clk = IO(Input(Clock()))
+  *   // Implement the virtual method to tell Chisel about this Clock value
+  *   override protected def implicitClock = clk
+  *   // Now we have a Clock to use in this RawModule
+  *   val reg = Reg(UInt(8.W))
+  * }
+  * }}}
+  */
+trait ImplicitClock { self: RawModule =>
```

Comment:
Does this need to be `RawModule` or can it be `BaseModule`?

Reply:
`RawModule` is the parent of all modules that are allowed to create hardware--it's not clear to me what being able to mix this into a `BlackBox` would accomplish. Note that this API only handles the implicit clock _within the module_, if we were to add a similar API for the outside ("implicitly hook up this port to the implicit clock when instantiated"), then I agree it should work for BlackBoxes as well.